ssz_serialization/merkleization.nim

```diff
@@ -443,7 +443,7 @@ func mixInLength*(root: Digest, length: int): Digest =
   dataLen[0..<8] = uint64(length).toBytesLE()
   mergeBranches(root, dataLen)
 
-func hash_tree_root*(x: auto): Digest {.gcsafe, raises: [Defect].}
+func hash_tree_root*(x: auto): Digest
```

Comment:
All forward declarations in Nim need explicit annotations. Otherwise, Nim has to assume the worst possible case (i.e. "gc unsafe", raises anything). 

The reason functions with bodies don't need these annotations is that the compiler is able to infer them from the body.

Reply:
Ah, thanks for that explanation! Would it make sense to copy the `{.gcsafe.}` to the implementation as well for consistency (even though it is implicitly inferred there), or should I just revert this back to the original?

And, for the `{.raises: [Defect].}`, is this still needed even though there is a `{.push raises: [Defect].}` at the top of the file?

Reply:
(reverting this section for now, but still interested in this, especially as this is one more of those things where the compiler does not complain if it is missing).

Reply:
Adding the same annotations to the body for consistency is a matter of style. I personally wouldn't add them, because the developer often needs to add the `gcsafe` pragma temporary while trying to figure out why the compiler is considering a deep hierarchy of calls as unsafe. People will often forget to delete these temporarily added pragma, so each time I see `gcsafe` I have to wonder if it was put there by accident and I'm tempted to spend (read waste) time investigating this.

Reply:
OK, it makes sense to keep the same style as before, then. While I do not see any compilation problems without either of the two annotations in this specific case, keeping them does not hurt (and does not have to be bundled with this PR).